Add basic CRUD for BotInstanceService

timothyb89 opened the pull request:
Related: #41456

This adds an early backend for the bot instance service, with resource definitions and basic CRUD operations.

Note that `SubmitHeartbeat` remains a stub, to be implemented in a follow up PR. There are no public methods to create bot instances, as this will be done internally by auth on bot join.

timothyb89 committed:
Add basic CRUD for BotInstanceService

This adds an early backend for the bot instance service, with resource
definitions and basic CRUD operations.

timothyb89 commented on lib/services/local/bot_instance.go:
```diff
@@ -0,0 +1,129 @@
+// matching the given (botName, instanceID) key and persists the patched
+// resource. It will make multiple attempts if a `CompareFailed` error is
+// raised.
+func (b *BotInstanceService) PatchBotInstance(
```
I've added the `Patch` implementation based on the one from `PresenceService`, though as @strideynet and I discussed we might want to split this out into a generic version on `ServiceWrapper` in a separate PR.

strideynet commented on api/proto/teleport/machineid/v1/bot_instance.proto:
```diff
@@ -0,0 +1,116 @@
+  string join_method = 6;
+  // Indicates whether `tbot` is running in one-shot mode.
+  bool one_shot = 7;
+  // In future iterations, additional information can be submitted here.
```
Figured Arch and OS may be useful to someone.

```suggestion
  // The architecture of the host that `tbot` is running on, determined by
  // runtime.GOARCH.
  string architecture = 8;
  // The OS of the host that `tbot` is running on, determined by runtime.GOOS.
  string os = 9;
  // In future iterations, additional information can be submitted here.
```

strideynet commented on api/proto/teleport/machineid/v1/bot_instance.proto:
```diff
@@ -0,0 +1,116 @@
+
+// BotInstanceStatus holds the status of a BotInstance.
+message BotInstanceStatus {
+  // The unique identifier for this bot.
+  string id = 1;
```
Not yet - this comment should really something more like "The unique identifier of this BotInstance"

strideynet commented on api/proto/teleport/machineid/v1/bot_instance.proto:
```diff
@@ -0,0 +1,116 @@
+message BotInstanceStatus {
+  // The unique identifier for this bot.
+  string id = 1;
+  // The name of the Bot that this instance is associated with.
+  string bot_name = 2;
```
Yeah - I think the plan is that BotInstance.Metadata.Name will either be `id` or `bot_name/id`. BotInstance's belong to a Bot, which is what's being indicated by `bot_name` here. We plan to store these in the backend with a key of `bot_name/id` so that we can easily find all BotInstances that belong to a bot.

timothyb89 replied:
Yes, I think the final `.metadata.name` of a bot instance will have to be `$botName/$instanceID`, leading to some tctl UX along the lines of `tctl get botinstance/foo/abcd1234` (I think there's precedent for that?)

timothyb89 replied:
I think I was off before and the `.metadata.name` will just be the instance ID. This is using the built in service prefixing functionality so the bot name is already included in the key and we don't want to include it twice.

timothyb89 commented on api/proto/teleport/machineid/v1/bot_instance.proto:
```diff
@@ -0,0 +1,116 @@
+
+// BotInstanceSpec contains fields
+message BotInstanceSpec {
+  // Empty as is not user configurable.
```
I have some uncertainty with how metadata will be handled on create. The notification service has some prior art here, either [mandating nil metadata](https://github.com/gravitational/teleport/blob/master/lib/services/local/notifications.go#L220) on the incoming resource and extracting the name (uuid) from a field in the spec, or generating the name on demand and [overwriting the `.metadata.name`](https://github.com/gravitational/teleport/blob/master/lib/services/local/notifications.go#L143).

I'd like to avoid exposing the `$botName/$instanceID` string formatting logic to API users, so it might make sense to include `bot_name` as part of `BotInstanceSpec` to allow for this? (and possibly also `id` depending on which component is actually generating the uuid)

strideynet replied:
Hm yeah - tbh - maybe we need to rethink if we even need to Create/Update at the gRPC layer ?

timothyb89 replied:
Per our offline discussion, I'll be moving `BotName` and `InstanceId` into the `Spec` field and having metadata generated on create. Since it's using the built in generic prefixing functionality that does mean we can't include the bot name in the instance's `.metadata.name`, so it'll just be the UUID. I think that's ultimately for the best though.

timothyb89 committed:
Update api/proto/teleport/machineid/v1/bot_instance.proto

Co-authored-by: Noah Stride <noah.stride@goteleport.com>

timothyb89 committed:
Further work on BotInstance boilerplate

Among other things, this adds event definitions, RBAC and presets,
and client definitions. This also includes a few proto changes,
namely moving bot name and instance ID from .Status to .Spec so
metadata can be uniformly generated.

This also adds tests for the backend service showing functional CRUD.

timothyb89 commented on api/proto/teleport/machineid/v1/bot_instance.proto:
```diff
@@ -0,0 +1,121 @@
+  // The desired expiration offset for this bot instance. The expiration will be
+  // calculated from the current time at creation or authentication plus this
+  // value. A nil `ttl` will not expire.
+  google.protobuf.Duration ttl = 3;
```
I've made some minor changes here, partly as discussed: `bot_name` and (now) `instance_id` are part of the initial spec so metadata can be generated from it, and the desired TTL is included now, so we can calculate a desired expiration, and update it appropriately on renew/rejoin/etc.

timothyb89 committed:
gRPC tests; fix filtering; fix import lints; misc cleanup

timothyb89 committed:
Consistently use `instance_id` in gRPC fields

timothyb89 committed:
Add tctl support for bot_instance resource

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/bot-instance-service

timothyb89 committed:
protos update

timothyb89 committed:
Various typo fixes

timothyb89 committed:
Fix more comment typos

timothyb89 committed:
More doc comments and minor test tweaks

timothyb89 commented:
I think this should be ready for review now.

An implementation note / question (?): I've opted not to add caching to the service at the moment. My intuition is that bot instances will be a write-many / read-few type of resource and I _think_ are not appropriate for caching, per the [resource guidelines RFD](https://github.com/gravitational/teleport/blob/master/rfd/0153-resource-guidelines.md#cache).

We've also opted to not include bot instance creation and will follow up with that in a separate PR shortly. Bot instances will be created internally by auth when bots authenticate so create functionality will not be exposed via gRPC or to tctl / etc. This admittedly does make manual testing a bit tricky.
